### PR TITLE
[REVIEW] Fix concurrent JSON reads crash

### PR DESCRIPTION
**Pull request by vuule**
Closes: #5934 

Root cause:
A temporary `unique_ptr` to a `mutable_table_device_view` object was being passed to `device_scalar`. The error only reproes if a concurrent thread/process changes the content at the temporary location -  no repro with a singlethreaded read.

Fix:
This PR contains a simple fix, and a few fixes for incorrect stream use in the JSON reader.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/io/json/reader_impl.cu [hunk 1]
```diff
@@ -156,9 +156,9 @@ std::unique_ptr<table> create_json_keys_info_table(
   info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT16), num_keys));
   info_columns.emplace_back(make_numeric_column(data_type(type_id::UINT32), num_keys));
   // Create a table out of these columns to pass them around more easily
-  auto info_table = std::make_unique<table>(std::move(info_columns));
-  rmm::device_scalar<mutable_table_device_view> info_table_mdv(
-    *mutable_table_device_view::create(info_table->mutable_view(), stream), stream);
+  auto info_table           = std::make_unique<table>(std::move(info_columns));
+  auto const info_table_mdv = mutable_table_device_view::create(info_table->mutable_view(), stream);
+  rmm::device_scalar<mutable_table_device_view> d_info_table_mdv(*info_table_mdv, stream);
```
Comment (jrhemstad) [inline cpp/src/io/json/reader_impl.cu:161]: Why is this creating a `device_scalar<mutable_table_device_view>`? That sounds really odd to me. `mutable_table_device_view` should be passed by value into kernels.

File: cpp/src/io/json/reader_impl.cu [hunk 2]
```diff
@@ -169,7 +169,7 @@ std::unique_ptr<table> create_json_keys_info_table(
                                          record_starts.data().get(),
                                          record_starts.size(),
                                          key_counter.data(),
-                                         info_table_mdv.data(),
+                                         d_info_table_mdv.data(),
```
Comment (jrhemstad) [inline cpp/src/io/json/reader_impl.cu:172]: Just pass it by value?  
```suggestion
                                         *d_info_table_mdv,
```

Comment (vuule) [reply cpp/src/io/json/reader_impl.cu:172]: @jrhemstad  It's an optional parameter, so I need to pass an optional-like thing (i.e. pointer). This is the best way I found to do this.

**Comment by vuule** [comment]
> I'll approve since this is a release blocker and I know cuIO code is going to undergo a refactor in the near future, but passing a view type by pointer is smelly.

Changed the parameter to `thrust::optional`.

**Comment by vuule** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Pull request merged**